### ISSUE TITLE
Event based damage calculator

### ISSUE THREAD

**Issue by raoanirudh** [gem/oq-engine#5339]
## Introduction
Heretofore the engine has provided five primary risk calculators: `classical_damage`, `classical_risk`, `scenario_damage`, `scenario_risk`, and `event_based_risk`. A stochastic `event_based_damage` calculator has been the critical missing piece. The rationale and specifications for building such an event based damage calculator are proposed in this issue.

## Rationale
The classical PSHA based damage/risk approach is useful if the user is mainly interested in estimating damage-state probabilities or loss exceedance probabilities within a time-window for individual assets. This approach is also conducive for the computation of time-averaged metrics such as the _average annual collapses_ (or 50-year collapse probabilities) and the _average annual loss_.

While the average annual loss values for different assets that lie within the same target accumulation zones (or any other arbitrary groups of assets that share a set of attributes) can be simply added to yield the aggregate average annual loss for the target accumulation zones (or for the groups of assets), the same doesn't hold for the loss exceedance curves or probabilistic loss maps outputs. For instance, simply adding up the 2,500-year return period losses for all of the assets within an accumulation zone (or any other arbitrary aggregation group) will not necessarily yield the 2,500-year return period loss for the zone (or group of assets). For computing almost all kinds of aggregated loss statistics for a distributed portfolio of assets, the `event_based_risk` calculator provides a much more viable solution. The `event_based_damage` calculator would serve a similar function, allowing users to compute **aggregated damage statistics** for a distributed portfolio of assets

In addition, the `event_based_damage` calculator would also enable the calculation of probabilistic damage metrics such as **damage-state exceedance curves** and **damage-state maps** for different return periods, which the `classical_damage` calculator does not support. Similar to the event based risk calculator, the `event_based_damage` calculator would be able to compute these metrics at both the asset level and also at the aggregated level for groups of assets sharing a set of attributes.

Finally, the central result computed by the `event_based_damage` calculator, viz the _asset-level event damage table_ – would serve as the starting input for estimating a wide set of **consequences** including but not limited to direct economic costs of repairing the damaged buildings, estimates of casualties, displaced households, shelter requirements, loss of use of essential facilities, amount of debris generated, costs associated with loss of function (e.g., business interruption losses), and repair and recovery times. While the computation of consequences would be coupled with the `event_based_damage` calculator, specifications for a general consequences calculator will be proposed in a separate issue and this issue focuses solely on the `event_based_damage` calculator.

## Specifications

**Note**: While these specifications are written primarily with the stochastic `event_based_damage` calculator in mind, notes are also provided for the `scenario_damage` calculator given that the internals of the scenario and event-based calculators have been on a path of convergence. Any changes necessary for the `event_based_damage` calculator should also be considered for the `scenario_damage` calculator.

### 1. Event set

The starting point for the `event_based_damage` calculator is typically an event set. This could be a stochastic event set generated for a time-based calculation, or a scenario event set for a scenario-based calculation. An events table establishes the associations between the ids of the events in the event set, the ids of the underlying ruptures, and the ids of the realisations representing the selected logic tree paths:

| `event_id` | `rup_id` | `rlz_id` |
|----------|--------|--------|
| 0        | 0      | 0      |
| 1        | 0      | 0      |
| 2        | 0      | 1      |
| 3        | 1      | 1      |
| 4        | 2      | 2      |
| ⋮        | ⋮      | ⋮      |
| 9320     | 515    | 7      |
| 9321     | 516    | 7      |

Information about the events, ruptures, and realisations is accessible to the engine at this stage, most typically this will be read from the datastore for a previously completed hazard calculation.

**Note**: Currently, the `scenario` calculator always produces only a single "deterministic" rupture, so the `rup_id` column in the events table for a scenario calculation contains only zeros. In principle this constraint could be relaxed in the future, allowing for the modelling and propagation of uncertainty in the rupture parameters.

**Note**: An alternative entry point for the `event_based_damage` calculator might be ground motion fields input by the user. This is a tricky use case that must be managed carefully and specifications will be proposed in a separate issue.

### 2. Ground motion fields (GMFs)

For each event, a set of ground motion fields or ground shaking intensity fields are simulated using the GSIM associated with the `rlz_id` for the event; one field for each intensity measure type (IMT) required for the calculation. The IMTs are either inferred from the set of fragility files listed in the job file, or specified explicitly using the `intensity_measure_types` parameter.

The sites at which the ground motion fields need to be computed can be specified in several ways, which are listed below in decreasing order of priority. I.e., if more than one of these conditions holds true, the one with the highest rank will take precedence and the rest will be ignored. This priority ranking is the same as the engine's current behaviour.

1. 	By using the parameter `sites`
2.	By providing a `sites_csv` file
3.	By specifying a `region` and `region_grid_spacing`
4.	By providing an `exposure_file` and specifying a `region_grid_spacing`: a `region` is created enveloping the asset locations in the exposure file, and this region is gridded with the specified spacing; furthermore, grid points with no assets in their vicinity are discarded
5.	By providing a `site_model_file` and specifying  a `region_grid_spacing`: a `region` is created enveloping the sites in the site model file, and this region is gridded with the specified spacing
6.	By providing a `site_model_file` without specifying a `region_grid_spacing`: the list of sites in the site model file are directly used
7.	By providing an ` exposure_file` without specifying a `region_grid_spacing`: the locations of the assets are directly used

At this stage the main calculated result is the GMF table, an example of which is shown below.

| `event_id` | `site_id` | `gmv_PGA`      | `gmv_SA(0.1)`  | `gmv_SA(0.3)`  | `gmv_SA(1.0)`  |
|----------|---------|--------------|--------------|--------------|--------------|
| 0        | 0       | 1.3779E-02 | 1.3788E-02 | 2.3977E-02 | 1.7175E-02 |
| 0        | 1       | 3.5265E-02 | 8.7311E-02 | 4.2909E-02 | 1.8889E-02 |
| 0        | 2       | 6.9670E-02 | 7.0575E-02 | 2.5058E-02 | 1.0375E-01 |
| ⋮        | ⋮       | ⋮            | ⋮            | ⋮            | ⋮            |
| 0     | 2064   | 9.2375E-01 | 8.4724E-01 | 5.0618E-01 | 1.2813E+00 |
| ⋮        | ⋮       | ⋮            | ⋮            | ⋮            | ⋮            |
| 9321     | 20734   | 1.9257E-01 | 2.7195E-01 | 5.2430E-01 | 7.3216E-01 |
| 9321     | 20735   | 8.5560E-01 | 1.5595E+00 | 1.9409E+00 | 6.3391E-01 |

To be able to manage the volume of data efficiently, the event-site combinations for which all of the ground motion values (GMVs) are either zero or fall below the `minimum_intensity` threshold are dropped from the table. The sites are stored in a site mesh table, which is simply a list of site_ids and the associated coordinates.

| `site_id` | `lon`        | `lat`      |
|---------|------------|----------|
| 0       | -123.57000 | 36.82000 |
| 1       | -123.57000 | 36.83000 |
| ⋮        | ⋮  | ⋮  |
| 2064       | -122.42000 | 37.76000 |
| ⋮        | ⋮  | ⋮  |
| 3735       | -121.00000 | 38.11000 |

### 3.1 Damage state sampling

For each event, for each asset in the exposure model, the GMVs at the hazard site associated with the asset and corresponding to the IMTs used by the fragility functions associated with the asset's taxonomy string, are used to calculate the damage state probabilities (currently named 'fractions' in the engine). In the example below, the exposure model comprises assets with the mandatory fields being only `asset_id`, `lon`, `lat`, `taxonomy`, and `number`. `zip_code` is an optional field that is used in this example for indicating target accumulation/aggregation zones.

| `asset_id` | `lon` | `lat`      | `taxonomy`  | `number`  | `zip_code`  |
|-----------|-------|--------|--------------|------------|------------|
| 0        | -122.42692       | 37.76441 | TAX1 | 6 | 94114 |
| ⋮        | ⋮       | ⋮            | ⋮            | ⋮            | ⋮            |
| 26,574        | ⋯   | ⋯ | ⋯ | ⋯ | ⋯ |

In this case, the asset with `asset_id = 0` is associated with the hazard site with `site_id = 2064` and is associated with a structural fragility function conditioned on SA(0.3) as the IMT. Thus for the event with `event_id = 0`, the relevant GMV for computing structural damage states for building units comprising this asset is 5.0618E-01g or 0.50618g. This value is used to get the damage state probabilities or 'fractions' for the asset from the structural fragility function, illustrated in the figure below. In this example, ```fractions = [0.115, 0.223, 0.306, 0.229, 0.127]``` represents occurrence probabilities for the ```[no_damage, slight, moderate, extensive, complete]``` damage states respectively.

**Note**: The fractions must always sum to 1.0 (to within a tolerance of [~1.49e-8](https://github.com/numpy/numpy/blob/v1.17.4/numpy/random/mtrand.pyx#L778-L781)). This should be guaranteed because the no damage probability is always calculated as 1.0 minus the sum of the probabilities of all other damage states. 

![damage-state-fractions](https://user-images.githubusercontent.com/1707414/69880257-d3d68800-12c9-11ea-89de-0809d3c3e55a.png)

Using these damage state probabilities, we sample damage states for each of the building units comprising the asset. Since the asset under consideration in the example has 6 building units (`number = 6`), one of five damage states for each building unit is sampled using `numpy.random.choice()`

```python
ds = numpy.random.choice(5, 6, p=[0.115, 0.223, 0.306, 0.229, 0.127])
>>> ds
array([0, 3, 4, 1, 4, 3])
```

We need to store only the total number of building units in each damage state for our entry in the asset-level event damage table, so we can count the instances of each damage state using `numpy.bincount()`

```python
ds_counts = numpy.bincount(ds)
>>> ds_counts
array([1, 1, 0, 2, 2])
```

### 3.2 Event damage table (EDT)

The full calculation would result in filling up the asset-level event damage table, which has max dimensions of `(num_events × num_assets × num_loss_types × num_damage_states)`.

```python
import numpy

rng = numpy.random.default_rng() # numpy ≥ 1.17 only

E = len(events)
A = len(assets)
L = len(crmodel.loss_types)
D = len(crmodel.damage_states)

event_damage_table = numpy.zeros(shape=[E, A, L, D], dtype=int)

for event in events:
    # Compute ground motion fields for event
    ⋮
    for asset in assets:
        n = asset["number"]
        for loss_type in loss_types:
            # Obtain ground motion value for asset and loss type
            ⋮
            # Obtain damage state fractions for asset and loss type
            ⋮
            damage_state_counts = numpy.bincount(rng.choice(D, n, p=fractions)) # numpy ≥ 1.17
            damage_state_counts = numpy.bincount(numpy.random.choice(D, n, p=fractions)) # numpy < 1.17
            event_damage_table[event_id, asset_id, loss_type] = damage_state_counts
```

In principle, the counts for the `no_damage` state could be dropped from the table, since these counts can always be obtained when needed simply by subtracting the sum of the counts for all other damage states from the total number of building units for the asset.

The uncompressed EDT, in a flattened form would thus contain results as in the following table:

| `event_id` | `asset_id` | `structural_ds1` | `structural_ds2` | `structural_ds3` | `structural_ds4` | `nonstructural_ds1` | ⋯ |
|----------|----------|----------------|----------------|----------------|----------------|-------------------|---|
| 0        | 0        | 1              | 0              | 2              | 2              | ⋯                 | ⋮ |
| ⋮        | ⋮        | ⋮              | ⋮              | ⋮              | ⋮              | ⋮                 | ⋮ |
| 0        | 26,574   | ⋯              | ⋯              | ⋯              | ⋯              | ⋯                 | ⋯ |
| ⋮        | ⋮        | ⋮              | ⋮              | ⋮              | ⋮              | ⋮                 | ⋮ |
| ⋮        | ⋮        | ⋮              | ⋮              | ⋮              | ⋮              | ⋮                 | ⋮ |
| 9321     | 26,574   | ⋯              | ⋯              | ⋯              | ⋯              | ⋯                 | ⋯ |

The event damage table is the primary output of the `event_based_damage` calculator. Other useful outputs such as average (annual) collapse counts, damage-state exceedance curves, and damage-state maps for given probabilities of exceedance within a certain `risk_investigation_time` or for given `return_periods` can be derived from the event damage table.

### 4 Average asset/aggregate damages
The average asset damages (`avg_damages`) output is analogous to the average asset losses output from the `event_based_risk` and `ebrisk` calculators. This output calculates the time-averaged number of building units in different damage states for each asset, for each loss category. This output is first calculated individually for each realisation: 

```python
for rlz in rlzs:
    event_ids = events_by_rlz[rlz]
    for asset_id in asset_ids:
        for loss_type in loss_types:
            for damage_state in damage_states:
                avg_damages[asset_id][rlz][loss_type][damage_state] = (
                    sum(event_damage_table[event_ids][asset_id][loss_type][damage_state])
                     * risk_investigation_time / eff_inv_time)
```
...where `eff_time` is the effective investigation time for the realisation,
i.e., `eff_time = (ses_per_logic_tree_path × investigation_time)`

The weighted mean and other statistics are then computed from the individual realisation results. Similar to the other two calculators, only the Average Asset Damages Statistics outputs will be exposed to the user.

Aggregate Damages (`agg_damages-rlz`) and Aggregate Damages Statistics (`agg_damages-mean`, `agg_damages-quantile`) are obtained by summing up the average asset damage results for the tag-combinations requested in the job file using the parameter `aggregate_by`, similar to the analogous loss results for `ebrisk`. Total Damages (`tot_damages-rlz`) and Total Damages Statistics (`tot_damages-mean`, `tot_damages-quantile`) are obtained by summing up the total damage results for the entire portfolio of assets.

### 5 Consequences and Event Consequence Table [WIP]
To be added

### 6 Consequences Exceedance Curves [WIP]
To be added

### Exposure files [WIP]
One of the key requirements for enabling the assignment of integral counts for the different damage states for each asset is that the number of building units for each asset in the exposure model **must** be integers. This will entail implementing a check for exposures to verify that all entries in the `number` field are integers. Because this will be a breaking change from the existing schema, we might consider upgrading the version number for exposure files from `nrml/0.5` to `nrml/0.6`, if that is still relevant. Jobs involving older exposure files that have non-integer counts in the `number` field will not be able to use the `event_based_damage` calculator or the new scenario damage calculator.

### Scenario damage
The current scenario damage calculator needs to be gradually phased out in favour of the updated calculator using the damage state sampling approach. While the old set of outputs would still be generated, the new calculator would also be able to generate asset-level event damage tables. The change of algorithm necessitates writing new test cases.

**Comment by raoanirudh** [comment]
The pending implementation of handling consequences tied to the asset-level event damage table and typical metrics for those consequences deserve a separate issue and specifications. Thus, closing this issue as implemented successfully in #4736 and #5802. A demo for the new calculator has been added in #5934 and relevant sections have been added to the engine manual in #6013. The `scenario_damage` and `event_based_damage` calculators have been unified in #6102.

**Comment by raoanirudh** [comment]
Reopening issue to add specs for the event consequence table, average (annual) consequences, and consequence exceedance curves
